### PR TITLE
feat(cli): introduce remote config for quickstart 

### PR DESCRIPTION
**Pull request by szalai1**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/cli/docker_check.py [hunk 1]
```diff
@@ -186,10 +169,37 @@ def to_exception(
...
+            container = gms[0]
+            image = container.attrs['Config']['Image']
+            return image.split(":")[1]
+        else:
+            return None
+
+def check_docker_quickstart(containers_ensure_exits: List[str], containers_required: List[str]) -> QuickstartStatus:
     container_statuses: List[DockerContainerStatus] = []
-
+    get_running_gms_version()
```
Comment (szalai1) [inline metadata-ingestion/src/datahub/cli/docker_check.py:202]:
```suggestion
```

File: docker/docker-compose.yml [hunk 2]
```diff
@@ -68,6 +68,8 @@ services:
 
   # This "container" is a workaround to pre-create search indices
   elasticsearch-setup:
+    labels:
+      setup_job: true
```
Comment (hsheth2) [inline docker/docker-compose.yml:72]: Would be nice to qualify this with `datahub`

```suggestion
      datahub_setup_job: true
```

File: docker/quickstart/quickstart_version_mapping.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  default:
+    composefile_git_ref: master
+    docker_tag: head
+  # v0.9.6 images contain security vulnerabilities 
+  v0.9.6: 
+    composefile_git_ref: v0.9.6.1
+    docker_tag: v0.9.6.1
+  stable: 
+    composefile_git_ref: master
+    docker_tag: head
```
Comment (hsheth2) [inline docker/quickstart/quickstart_version_mapping.yaml:18]: should stable be pointing at head/master?

File: metadata-ingestion/src/datahub/cli/docker_cli.py [hunk 4]
```diff
@@ -803,28 +797,43 @@ def quickstart(
     )
 
 
+def get_docker_compose_base_url(version_tag: Optional[str]) -> str:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/docker_cli.py:800]:
```suggestion
def get_docker_compose_base_url(version_tag: str) -> str:
```

File: metadata-ingestion/tests/cli/test_quickstart_version_mapping.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    execution_plan = example_version_mapper.get_quickstart_execution_plan("v1.0.0")
+    expected = QuickstartExecutionPlan(
+        docker_tag="v1.0.0",
+        composefile_git_ref="v1.0.0",
+    )
+    assert execution_plan == expected
+
+def test_quickstart_version_config_default():
+    execution_plan = example_version_mapper.get_quickstart_execution_plan("v2.0.0")
+    excepted = QuickstartExecutionPlan(
```
Comment (hsheth2) [inline metadata-ingestion/tests/cli/test_quickstart_version_mapping.py:35]:
```suggestion
    expected = QuickstartExecutionPlan(
```

File: docker/quickstart/quickstart_version_mapping.yaml (hunk 3 above)
Comment (szalai1) [reply docker/quickstart/quickstart_version_mapping.yaml]: in an ideal world yes. changed this, if not defined fetches the latest release tag from github

File: docker/docker-compose-without-neo4j.override.yml [hunk 6]
```diff
@@ -15,7 +15,7 @@ services:
 
   mysql-setup:
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/docker-compose-without-neo4j.override.yml:18]:
```suggestion
      datahub_setup_job: true
```

File: docker/docker-compose.override.yml [hunk 7]
```diff
@@ -16,7 +16,7 @@ services:
 
   mysql-setup:
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/docker-compose.override.yml:19]:
```suggestion
      datahub_setup_job: true
```

File: docker/docker-compose.yml [hunk 8]
```diff
@@ -69,7 +69,7 @@ services:
   # This "container" is a workaround to pre-create search indices
   elasticsearch-setup:
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/docker-compose.yml:72]:
```suggestion
      datahub_setup_job: true
```

File: docker/docker-compose.yml [hunk 9]
```diff
@@ -121,7 +121,7 @@ services:
   # explicitly wait for this container
   kafka-setup:
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/docker-compose.yml:124]:
```suggestion
      datahub_setup_job: true
```

File: docker/docker-compose.yml [hunk 10]
```diff
@@ -135,7 +135,7 @@ services:
 
   datahub-upgrade:
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/docker-compose.yml:138]:
```suggestion
      datahub_setup_job: true
```

File: docker/quickstart/docker-compose-m1.quickstart.yml [hunk 11]
```diff
@@ -128,7 +128,7 @@ services:
     hostname: datahub-upgrade
     image: ${DATAHUB_UPGRADE_IMAGE:-acryldata/datahub-upgrade}:${DATAHUB_VERSION:-head}
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/quickstart/docker-compose-m1.quickstart.yml:131]:
```suggestion
      datahub_setup_job: true
```

File: docker/quickstart/docker-compose-m1.quickstart.yml [hunk 12]
```diff
@@ -160,7 +160,7 @@ services:
     hostname: elasticsearch-setup
     image: ${DATAHUB_ELASTIC_SETUP_IMAGE:-linkedin/datahub-elasticsearch-setup}:${DATAHUB_VERSION:-head}
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/quickstart/docker-compose-m1.quickstart.yml:163]:
```suggestion
      datahub_setup_job: true
```

File: docker/quickstart/docker-compose-m1.quickstart.yml [hunk 13]
```diff
@@ -173,7 +173,7 @@ services:
     hostname: kafka-setup
     image: ${DATAHUB_KAFKA_SETUP_IMAGE:-linkedin/datahub-kafka-setup}:${DATAHUB_VERSION:-head}
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/quickstart/docker-compose-m1.quickstart.yml:176]:
```suggestion
      datahub_setup_job: true
```

File: docker/quickstart/docker-compose.quickstart.yml [hunk 14]
```diff
@@ -126,7 +126,7 @@ services:
     hostname: datahub-upgrade
     image: ${DATAHUB_UPGRADE_IMAGE:-acryldata/datahub-upgrade}:${DATAHUB_VERSION:-head}
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/quickstart/docker-compose.quickstart.yml:129]:
```suggestion
      datahub_setup_job: true
```

File: docker/quickstart/docker-compose.quickstart.yml [hunk 15]
```diff
@@ -157,7 +157,7 @@ services:
     hostname: elasticsearch-setup
     image: ${DATAHUB_ELASTIC_SETUP_IMAGE:-linkedin/datahub-elasticsearch-setup}:${DATAHUB_VERSION:-head}
     labels:
-      datahub_setup_job: true
+      datahub_setup_job: true
```
Comment (szalai1) [inline docker/quickstart/docker-compose.quickstart.yml:160]:
```suggestion
      datahub_setup_job: true
```

File: metadata-ingestion/src/datahub/cli/docker_check.py [hunk 16]
```diff
@@ -189,28 +157,32 @@ def to_exception(
...
         containers = client.containers.list(
             all=True,
             filters=DATAHUB_COMPOSE_PROJECT_FILTER,
         )
+        if len(containers) == 0:
+            return QuickstartStatus([])
+
+        # load the expected containers from the docker-compose file
+        config_file = containers[0].labels.get(
+            "com.docker.compose.project.config_files"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/docker_check.py:170]: what happens if they use multiple docker compose files

e.g. `docker-compose -p datahub -f compose1.yml -f compose2.yml`?

File: metadata-ingestion/src/datahub/cli/docker_cli.py [hunk 17]
```diff
@@ -916,7 +926,7 @@ def ingest_sample_data(path: Optional[str], token: Optional[str]) -> None:
     status = check_docker_quickstart()
     if not status.is_ok():
         raise status.to_exception(
-            header="Docker is not ready:",
+            header="GMS is not healthy:",
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/docker_cli.py:929]: revert this change

File: metadata-ingestion/src/datahub/cli/docker_check.py [hunk 18]
```diff
@@ -189,28 +157,32 @@ def to_exception(
...
+        all_containers = []
+        with open(config_file, "r") as config_file:
+            all_containers = yaml.safe_load(config_file).get("services", {}).keys()
 
+        existing_containers = set()
         # Check that the containers are running and healthy.
         container: docker.models.containers.Container
         for container in containers:
-            name = container.name
+            name = container.labels.get("com.docker.compose.service", container.name)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/docker_check.py:180]: what's the difference between compose.service and container.name?

File: metadata-ingestion/src/datahub/cli/docker_check.py [hunk 19]
```diff
@@ -189,28 +157,32 @@ def to_exception(
...
             status = ContainerStatus.OK
-
-            if container.name not in (
-                REQUIRED_CONTAINERS + CONTAINERS_TO_CHECK_IF_PRESENT
-            ):
-                # Ignores things like "datahub-frontend" which are no longer used.
-                # This way, we only check required containers like "datahub-frontend-react"
-                # even if there are some old containers lying around.
+            if name not in all_containers:
                 continue
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/docker_check.py:184]: just for clarity, keeping the comment around to explain how this filters out old containers would be useful

File: metadata-ingestion/src/datahub/cli/docker_check.py [hunk 20]
```diff
@@ -189,28 +157,32 @@ def to_exception(
...
+        if len(containers) == 0:
+            return QuickstartStatus([])
+
+        # load the expected containers from the docker-compose file
+        config_file = containers[0].labels.get(
+            "com.docker.compose.project.config_files"
+        )
+        all_containers = []
+        with open(config_file, "r") as config_file:
+            all_containers = yaml.safe_load(config_file).get("services", {}).keys()
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/docker_check.py:174]: do we know that the file exists? if not, maybe we can fall back to checking all containers in the project

File: metadata-ingestion/src/datahub/cli/quickstart_versioning.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def parse_version(version: str) -> Tuple[int, int, int, int]:
+    """
+    Parses a version string into a tuple of integers.
+    :param version: The version string to parse.
+    :return: A tuple of integers representing the version.
+    """
+    version = re.sub(r"v", "", version)
+    parsed_version = tuple(map(int, version.split(".")))
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/quickstart_versioning.py:37]: let's use packaging.version.parse()

the object it returns also natively supports comparison

File: metadata-ingestion/src/datahub/cli/quickstart_versioning.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+        # if the version is older than v0.10.1, it doesn't contain the setup job labels and the
+        # the checks will fail, so in those cases we pick the composefile from v0.10.1 which contains
+        # the setup job labels
+        if is_it_a_version(result.composefile_git_ref):
+            if compare_versions(
+                parse_version("v0.10.1"), parse_version(result.composefile_git_ref)
+            ):
+                # The merge commit where the labels were added
+                # https://github.com/datahub-project/datahub/pull/7473
+                result.composefile_git_ref = "quickstart-stability"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/cli/quickstart_versioning.py:132]: can we use the exact commit hash instead, since it's possible someone deletes this branch accidentally

File: metadata-ingestion/tests/cli/test_quickstart_version_mapping.py [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+from datahub.cli.quickstart_versioning import QuickstartVersionMappingConfig, QuickstartExecutionPlan
+
+example_version_mapper = QuickstartVersionMappingConfig.parse_obj({
+    "quickstart_version_map": {
+        "default": {
+            "composefile_git_ref": "master",
+            "docker_tag": "latest"
+        },
+        "v0.9.6": {
+            "composefile_git_ref": "v0.9.6.1", # this will be ovewritten by the cli
```
Comment (hsheth2) [inline metadata-ingestion/tests/cli/test_quickstart_version_mapping.py:11]:
```suggestion
            "composefile_git_ref": "v0.9.6.1", # this will be overwritten by the cli
```

File: metadata-ingestion/tests/cli/test_quickstart_version_mapping.py [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+
+def test_quickstart_forced_not_a_version_tag():
+    """
+    If the user specifies a version that is not in the version mapping, we should still use the default version for checking.
+
+    This is because the user may be using a version of datahub that is not in the version mapping. Which is most likely
+    a recent change, otherwise it should be on an older version of datahub.
+    """
+    execution_plan = example_version_mapper.get_quickstart_execution_plan("NOT A VERSION")
+    excepted = QuickstartExecutionPlan(
```
Comment (hsheth2) [inline metadata-ingestion/tests/cli/test_quickstart_version_mapping.py:71]:
```suggestion
    expected = QuickstartExecutionPlan(
```

File: metadata-ingestion/tests/cli/test_quickstart_version_mapping.py [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+        docker_tag="NOT A VERSION",
+        composefile_git_ref="NOT A VERSION",
+    )
+    assert execution_plan == excepted
+
+def test_quickstart_get_older_version():
+
+    execution_plan = example_version_mapper.get_quickstart_execution_plan("v0.9.6")
+    excepted = QuickstartExecutionPlan(
```
Comment (hsheth2) [inline metadata-ingestion/tests/cli/test_quickstart_version_mapping.py:80]:
```suggestion
    expected = QuickstartExecutionPlan(
```

**Review by hsheth2 — CHANGES_REQUESTED** [review]
mostly looks good!

File: metadata-ingestion/src/datahub/cli/docker_check.py (hunk 18 above)
Comment (szalai1) [reply metadata-ingestion/src/datahub/cli/docker_check.py:180]: service name is how you call it in the composefile, the container name can be different. e.g actions sometimes gets a `-1` postfix.

File: metadata-ingestion/src/datahub/cli/docker_check.py (hunk 20 above)
Comment (szalai1) [reply metadata-ingestion/src/datahub/cli/docker_check.py]: it should exists as compose needs a file to start the services. if not, than something deleted it after start.  without the file we have no ways to determine which one belongs to datahub which not. (without using the mapping that we agreed on dumping)

File: metadata-ingestion/src/datahub/cli/docker_check.py (hunk 16 above)
Comment (szalai1) [reply metadata-ingestion/src/datahub/cli/docker_check.py:170]: address this.

File: docker/quickstart/quickstart_version_mapping.yaml [hunk 26]
```diff
@@ -0,0 +1,18 @@
...
+# quickstart_mapping maps broken releases to fixed ones
+quickstart_version_map:
+  # default key is mandatory and is used if no version is specified
+  default:
+    composefile_git_ref: master
+    docker_tag: head
+  # v0.9.6 images contain security vulnerabilities 
+  v0.9.6: 
+    composefile_git_ref: v0.9.6.1
+    docker_tag: v0.9.6.1
```
[the remark this answers is not in the document]
Comment (szalai1) [reply docker/quickstart/quickstart_version_mapping.yaml:14]: compose version: <the commit of this PR's merge>
image version: v0.9.5 

only exception needs to be listed.

File: metadata-ingestion/src/datahub/cli/docker_check.py (hunk 18 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/cli/docker_check.py:181]: ah makes sense

File: metadata-ingestion/src/datahub/cli/docker_check.py (hunk 20 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/cli/docker_check.py]: ok this should be fine then

**Review by hsheth2 — APPROVED** [review]
LGTM

I think there's one follow-up of using the commit hash instead of the branch name after merging, but that has to be done in a separate PR
